source/extensions/transport_sockets/tls/ssl_socket.h

```diff
@@ -77,11 +86,13 @@ class SslSocketInfo : public Envoy::Ssl::ConnectionInfo {
   std::string ciphersuiteString() const override;
   const std::string& tlsVersion() const override;
   absl::optional<std::string> x509Extension(absl::string_view extension_name) const override;
+  Network::PostIoAction doHandshake(Ssl::SocketState& state) override;
   SSL* ssl() const { return ssl_.get(); }
```

Comment:
Should we restrict access to the SSL* prior to handshake completing?

Reply:
Yes, I think so. It seems like a matter for the next PR, but my plan is (a) the state accessor method returns HandshakeOngoing, a new state enum value indicating that the handshake is ongoing, and (b) when state == HandshakeOngoing, ssl() might return nullptr.

Reply:
Sounds good.  The move of state_ to this object should make that future change possible.